src/HeapDump/GCHeapDumper.cs

```diff
@@ -928,15 +932,15 @@ private void DumpDotNetHeapDataWorker(DataTarget dataTarget, ClrRuntime[] runtim
             // ClrMD returns UOH segments with seg.Generation2.Start == seg.Start and seg.Generation2.End == seg.End
             // To make it easy to determine the real generation of the object, augmenting the object with Gen3End and
             // Gen4End as follows such that DotNetHeapInfo.GenerationFor works.
-            if (seg.IsPinnedObjectSegment)
+            if (seg.Kind == GCSegmentKind.Pinned)
```

Comment:
While not something you should tackle here, this set of `if` statements are questionable with the new GC segments feature.

I guess this all works correctly, but this takes a real dependency on the ordering of segment chunks.  Things don't really work that way when GC segments are enabled.

Reply:
@cshung, was this part of the area that you looked at when updating TraceEvent to support regions?

Reply:
Yes, I did update ClrMD and HeapDump to support regions.
The PRs are https://github.com/microsoft/clrmd/pull/1035 and https://github.com/microsoft/perfview/pull/1674 respectively.
I am not sure I understand what might have gone wrong here, @leculver can you elaborate?

Reply:
There's not anything in particular _wrong_ here.  I'm only flagging to check whether this particular code was reviewed and looks correct for GC_REGIONS.  Since you reviewed this and submitted #1674, I guess we already covered perfview.  This code just looks _odd_ to me in terms of how it handles segments/regions at all (the original code, not your updates).

In short, if it's already been reviewed from a GC_REGIONS perspective and you are good with it, then nothing to do.  I'm just double checking. :)

Reply:
Thanks for confirming. Here is how I see it.

In the old days, a segment could have multiple generations in it, so it makes sense to start with searching a containing region, and then check if it lands before gen2 ends, if it does, it is gen2, otherwise, check if it lands before gen 1, and so on. That's a pretty simple linear search.

Then it comes the large object heap, we could have check a flag and declare it is gen 3, but instead, PerfView chose to add a fake Gen3 end to the all segments, so that there is a fake gen 3 portion in the ephemeral segments as well, of length 0, starts and ends at the very beginning of the ephemeral segment (and also symetrically, add the fake gen 2/gen 1/gen 0 portions) that start and ends at the end of the large object segment, that will fit the linear search algorithm.

```txt
Under segments, an ephemeral segment looks like this:

|222222|11111|000|
|      |     |   +-------> Gen 0 Ends
|      |     +-----------> Gen 0 Start
|      |     +-----------> Gen 1 Ends
|      +-----------------> Gen 1 Start
|      +-----------------> Gen 2 Ends
+------------------------> Gen 2 Start
+------------------------> Gen 3 End
+------------------------> Gen 3 Start

Under regions, an ephemeral segment looks like this:

|1111111111111111|
|                +------> Gen 0 Ends
|                +------> Gen 0 Start
|                +------> Gen 1 Ends
+-----------------------> Gen 1 Start
+-----------------------> Gen 2 End
+-----------------------> Gen 2 Start
+-----------------------> Gen 3 End
+-----------------------> Gen 3 Start
```

My change to POH and regions simply follow suit. For regions, there is only one generation per region, so in fact all other generations in the same region is fake. 

I agree It is _odd_, the whole thing is bent backward to fit the original linear search algorithm, but I think it should work. In particular, it might look like there is some dependency on the ordering because of the linear search, but in fact, it doesn't, all that linear search does is going through the fakes. There will be exactly one region containing the object, and for each region, there will only be exactly one generation with a non zero length.

Reply:
Thanks for taking a look!